From 4fc448c64432943a7817f2c6ed2103eb4e8b25b2 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Sun, 19 Jan 2025 12:53:05 +0800
Subject: [PATCH 178/301] AOSCOS: drm: loongson: add ls7a2000_support module
 parameter

The loongson DRM driver does not implement userspace acceleration,
resulting in a lack of OpenGL/3D acceleration support. Loongson Technology
supplies a LoongGPU driver package which ships with a more complete
feature set. Users of 7A2000-based graphics are likely to get a better
experience out of this driver package instead.

Disable loongson driver probing by default for LS7A2000/2K2000 devices.

For those who feel adventurous, use the `loongson.ls7a2000_support=1'
parameter to enable this driver.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Link: https://t.me/c/1109254909/723685
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/gpu/drm/loongson/loongson_module.c | 4 ++++
 drivers/gpu/drm/loongson/loongson_module.h | 1 +
 drivers/gpu/drm/loongson/lsdc_drv.c        | 7 +++++++
 3 files changed, 12 insertions(+)

diff --git a/drivers/gpu/drm/loongson/loongson_module.c b/drivers/gpu/drm/loongson/loongson_module.c
index 89915a33ec71..99e5fef3f2d9 100644
--- a/drivers/gpu/drm/loongson/loongson_module.c
+++ b/drivers/gpu/drm/loongson/loongson_module.c
@@ -21,6 +21,10 @@ int loongson_ls7a1000_support = 0;
 MODULE_PARM_DESC(ls7a1000_support, "7A1000 support (1 = enabled, 0 = disabled (default))");
 module_param_named(ls7a1000_support, loongson_ls7a1000_support, int, 0400);
 
+int loongson_ls7a2000_support = 0;
+MODULE_PARM_DESC(ls7a2000_support, "7A2000 support (1 = enabled, 0 = disabled (default))");
+module_param_named(ls7a2000_support, loongson_ls7a2000_support, int, 0400);
+
 static int __init loongson_module_init(void)
 {
 	if (!loongson_modeset || video_firmware_drivers_only())
diff --git a/drivers/gpu/drm/loongson/loongson_module.h b/drivers/gpu/drm/loongson/loongson_module.h
index 0dcc9646d2ff..183a8a058660 100644
--- a/drivers/gpu/drm/loongson/loongson_module.h
+++ b/drivers/gpu/drm/loongson/loongson_module.h
@@ -8,6 +8,7 @@
 
 extern int loongson_vblank;
 extern int loongson_ls7a1000_support;
+extern int loongson_ls7a2000_support;
 extern struct pci_driver lsdc_pci_driver;
 
 #endif
diff --git a/drivers/gpu/drm/loongson/lsdc_drv.c b/drivers/gpu/drm/loongson/lsdc_drv.c
index ada66e809f36..31d187fe3d24 100644
--- a/drivers/gpu/drm/loongson/lsdc_drv.c
+++ b/drivers/gpu/drm/loongson/lsdc_drv.c
@@ -273,6 +273,13 @@ static int lsdc_pci_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 		return -ENODEV;
 	}
 
+	if (to_loongson_gfx(descp)->chip_id == CHIP_LS7A2000 &&
+	    !loongson_ls7a2000_support) {
+		dev_info(&pdev->dev,
+			"LS7A2000 support disabled or not enabled via module parameter\n");
+		return -ENODEV;
+	}
+
 	if (IS_ERR_OR_NULL(descp))
 		return -ENODEV;
 
-- 
2.48.1

